### ISSUE TITLE
Object Oriented in Manim Results in AttributeError: 'Object' object has no attribute 'foreground_mobjects'

### ISSUE THREAD

**Issue by mafifi** [ManimCommunity/manim#3291]
## Description of bug / unexpected behavior
I am trying to create a Sudoku grid class in Manim. If you're unfamiliar with Sudoku grids, they're a grid of boxes (9 of them), and each of them are a grid of cells (9 of them again). A perfect opportunity for OO if there ever was one.

So I have three classes,

- SudokuGrid
- Boxes
- Cells

Which necessitates overriding `__init__` to have member variables in their classes.
And whenever you override `__init__` manim just falls apart with the error in the subject line.
No attempt by me to move subsets of the code between `__init__` and `construct` has allowed me to overcome this issue.

See below for sample code and how to reproduce the issue.

## Expected behavior
This may be my lack of imagination, but I expect to be able to create OO objects that allow me to abstract complexity rather than have to manage all 91 boxes of a Sudoku grid individually every time I wish to animate one. Grouping them into a vgroup works of course if you only have one grid, but how do I then address the cells and populate their text or add more complex animations?

## How to reproduce the issue
<details><summary>Code for reproducing the problem</summary>

```py
class SudokuGrid(Scene):
	def __init__(self, gridStrokeWidth = 10, gridStrokeColor = WHITE, \
		boxStrokeWidth = 5, boxStrokeColor = WHITE, \
		cellStrokeWidth = 0.7, cellStrokeColor = WHITE, cellTextColor = WHITE):

		self.Border = Square(gridStrokeWidth, color = gridStrokeColor).scale(3)
		self.add(self.Border)
		self.Boxes = []
		for i in range(1, 4):
			for j in range(1, 4):
				x = i * 3 + j
				self.Boxes[x] = Box(boxStrokeWidth = boxStrokeWidth, \
					boxStrokeColor = boxStrokeColor, \
					cellStrokeWidth = cellStrokeWidth, \
					cellStrokeColor = cellStrokeColor, \
					cellTextColor = cellTextColor)
				self.Boxes[x].shift([(j -2) * 2, (i -2) * 2, 0])

	def construct(self):
		self.add(self.Border)
		for i in range(1, 4):
			for j in range(1, 4):
				x = i * 3 + j
				self.add(self.Boxes[x])
```

</details>

## Logs
<details><summary>Terminal output</summary>

```
File [~/.../.venv/lib/python3.11/site-packages/manim/scene/scene.py:223](https://file+.vscode-resource.vscode-cdn.net/Users/afifim/SudokuSleuth/~/SudokuSleuth/.venv/lib/python3.11/site-packages/manim/scene/scene.py:223), in Scene.render(self, preview)
    221 self.setup()
    222 try:
...
--> 468     mobjects = [*mobjects, *self.foreground_mobjects]
    469     self.restructure_mobjects(to_remove=mobjects)
    470     self.mobjects += mobjects

AttributeError: 'SudokuGrid' object has no attribute 'foreground_mobjects'
```

</details>

## System specifications

<details><summary>System Details</summary>

- OS: macOS 13.4.1 (Ventura)
- RAM: 32 GBs
- Python version (`python/py/python3 --version`): Python 3.11.4
- Installed modules (provide output from `pip list`):
```
Package                Version
---------------------- ----------
altgraph               0.17.2
certifi                2022.12.7
charset-normalizer     2.1.1
click                  8.1.4
click-default-group    1.2.2
cloup                  0.13.1
colour                 0.1.5
Cython                 0.29.36
decorator              5.1.1
filelock               3.9.0
future                 0.18.2
glcontext              2.3.7
huggingface-hub        0.11.1
idna                   3.4
isosurfaces            0.1.0
macholib               1.15.2
manim                  0.17.3
ManimPango             0.4.3
mapbox-earcut          1.0.1
markdown-it-py         3.0.0
mdurl                  0.1.2
moderngl               5.8.2
moderngl-window        2.4.4
multipledispatch       1.0.0
networkx               2.8.8
numpy                  1.24.1
packaging              22.0
Pillow                 9.5.0
pip                    21.2.4
pycairo                1.24.0
pydub                  0.25.1
pyglet                 2.0.8
Pygments               2.15.1
pyobjc-core            9.2
pyobjc-framework-Cocoa 9.2
pyrr                   0.10.3
PyYAML                 6.0
regex                  2022.10.31
requests               2.28.1
rich                   13.4.2
scipy                  1.11.1
screeninfo             0.8.1
setuptools             58.0.4
six                    1.15.0
skia-pathops           0.7.4
srt                    3.5.3
svgelements            1.9.5
tokenizers             0.13.2
tqdm                   4.64.1
transformers           4.25.1
typing_extensions      4.4.0
urllib3                1.26.13
watchdog               2.3.1
wheel                  0.37.0
```
</details>

<details><summary>FFMPEG</summary>

Output of `ffmpeg -version`:

```
ffmpeg version 6.0 Copyright (c) 2000-2023 the FFmpeg developers
built with Apple clang version 14.0.3 (clang-1403.0.22.14.1)
configuration: --prefix=/opt/homebrew/Cellar/ffmpeg/6.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-libjack --disable-indev=jack --enable-videotoolbox --enable-audiotoolbox --enable-neon
libavutil      58.  2.100 / 58.  2.100
libavcodec     60.  3.100 / 60.  3.100
libavformat    60.  3.100 / 60.  3.100
libavdevice    60.  1.100 / 60.  1.100
libavfilter     9.  3.100 /  9.  3.100
libswscale      7.  1.100 /  7.  1.100
libswresample   4. 10.100 /  4. 10.100
libpostproc    57.  1.100 / 57.  1.100
```
</details>

**Comment by uwezi** [comment]
The better (and correct) way would probably to create a new object (based on `Matrix` or `VMobject`) and use it in a regular scene than creating a whole new `Scene`-type object.
Since `Scene` is the object in your code which does all the rendering magic, you will easily break important stuff.

Above all you should always call the `super.__init__` of your inherited object and you should also make sure that your `__init__` accepts all parameters with which the parent object would be called. See the documentation for `Scene` and the corresponding source code of the original `Scene`'s `__init__`
https://docs.manim.community/en/stable/reference/manim.scene.scene.Scene.html

**Comment by TimeTravelPenguin** [comment]
> The better (and correct) way would probably to create a new object (based on `Matrix` or `VMobject`) and use it in a regular scene than creating a whole new `Scene`-type object.

I wanted to add to this to agree. In software, you have to consider composition vs inheritance. Does it makes sense to say that a "`SudokuGrid` *is a* `Scene`"? If not, then inheritance is not the way to go.

Make an object, that encapsulates what a `SudokuGrid` is, and write methods that allow you to interact with it. Eg, maybe a method to take a coordinate to update a particular entry in the grid. Then, you can use that object in *any* scene. It is more flexible, and you will save yourself rewriting a whole lot of code :)

**Comment by mafifi** [comment]
Thank you both, immensely helpful comments.
In hindsight it's obvious that I've mangled the parent object given I am not passing any of the parameters to `super()`, so unsurprisingly it's calling me out (just too used to statically checked languages compared to Python).

I did consider using composition over inheritance, you're correct in that it may be the way to go.
Will repost my revised working code once I have it in case it helps others.

Thanks again.
